### PR TITLE
Create hooks subcommand

### PR DESCRIPTION
**Pull request by russellkan**
Done:
- Added hook files
- Created `hooks` subcommand
- Created test for installation of hooks
- Created test for using hooks

To do:
- Add capability to process branch names starting with `feature/issue#-...`
- Processing edge case where branch name contains a non-issue numeral
- Implement installation of hooks over pre-existing hooks (perhaps by renaming existing hooks then installing new hooks)

### CODE REVIEW

File: tests/hook_test.py [hunk 1]
```diff
@@ -46,22 +46,25 @@ def prepare_branch(tmp_repo, branch_name):
 def test_single_issue(tmp_repo):
     prepare_branch(tmp_repo, '10-sample-issue')
 
-    assert subprocess.check_output(['git', 'show', '-s', '--format=%B'],
-        encoding='utf-8') == "Fix some issue\n\nIssue #10\n\n"
+    assert subprocess.check_output(
```
Comment (johndgiese) [inline tests/hook_test.py]: Perhaps a function, which would provide a nice name, would make these lines clearer?  E.g.,

```
def latest_git_message():
    return subprocess.check_output(['git', 'show', '-s', '--format=%B'], encoding='utf-8')

def test_single_issue(tmp_repo):
    prepare_branch(tmp_repo, '10-sample-issue')
    assert latest_git_message() == "Fix some issue\n\nIssue #10\n\n"
```

File: README.md [hunk 2]
```diff
@@ -55,6 +55,7 @@ RDM integrates tightly into modern software development workflows.  Essentially,
 - PyYAML
 - Pandoc and Latex (optional, required for PDFs)
 - Reportlab and Svglib (optional, required to include SVGs in PDFs)
+- GitPython (optional, required to test hooks)
```
Comment (johndgiese) [inline README.md]: I would split this out into a separate "Development Dependencies" section, or even just remove it.

We don't want potential users to be scared by the long dependency list.  And, since only developers working on `rdm` itself will need `GitPython` we can safely remove it from here.

File: rdm/cli.py [hunk 3]
```diff
@@ -25,13 +25,36 @@ def cli(raw_arguments):
...
 
 def init(output_directory):
     init_directory = pkg_resources.resource_filename(__name__, 'init')
     shutil.copytree(init_directory, output_directory)
 
 
+def install_hooks(dest=None):
+    hooks_source = pkg_resources.resource_filename(__name__, 'hooks')
+    if dest is None:
+        root = subprocess.check_output(['git', 'rev-parse', '--show-toplevel'])
```
Comment (johndgiese) [inline rdm/cli.py]: First of all: its great that you did not re-use the variable names here, and instead us `root` and `root_str`.

Second: In this situation, I tend to like making the longer-lived name (in this case `root`) take the nicer form.

For example, `root_as_bytes` and `root` instead of `root` and `root_str`.

I like this better because:

1. It indicates which variable is more important
2. It allows people reading the code to ignore the fact that `root` is a string, later on, since it is no longer relevant after the conversion.

Does this make sense?  please push back if you don't think it does

That said, both variable's scope is so short it doesn't really matter at the moment---but who knows how the code will evolve in the future.

File: rdm/cli.py [hunk 4]
```diff
@@ -25,13 +25,36 @@ def cli(raw_arguments):
...
+    copyhooks(hooks_source, dest)
+    print('Successfully installed hooks in {}'.format(dest))
+
+
+def copyhooks(source, dest):
+    if not os.path.exists(dest):
+        os.makedirs(dest)
+    for item in os.listdir(source):
+        src = os.path.join(source, item)
+        dst = os.path.join(dest, item)
```
Comment (johndgiese) [inline rdm/cli.py]: I think it is bad to use `dest` and `dst` in the same function.  Maybe find better names for both `src` and `dst`.  E.g. `item_src` and `item_dest`?

File: rdm/cli.py [hunk 5]
```diff
@@ -25,13 +25,36 @@ def cli(raw_arguments):
...
+    hooks_source = pkg_resources.resource_filename(__name__, 'hooks')
+    if dest is None:
+        root = subprocess.check_output(['git', 'rev-parse', '--show-toplevel'])
+        root_str = root.strip().decode('ascii')
+        dest = root_str + '/.git/hooks'
+    copyhooks(hooks_source, dest)
+    print('Successfully installed hooks in {}'.format(dest))
+
+
+def copyhooks(source, dest):
```
Comment (johndgiese) [inline rdm/cli.py]: The name of this function seems to specific for what it actually does.  Perhaps `copy_directory` or something like it would be more appropriate?

File: tests/hook_test.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+def show_commit_message():
+    return str(subprocess.check_output(
+        ['git', 'show', '-s', '--format=%B'],
+    ), 'utf-8')
+
+
+def test_single_issue(tmp_repo):
+    prepare_branch(tmp_repo, '10-sample-issue')
+
+    assert show_commit_message() == "Fix some issue\n\nIssue #10\n\n"
```
Comment (johndgiese) [inline tests/hook_test.py]: IMO the new lines in these tests don't make the code easier to read, but do make the file longer.

File: tests/hook_test.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def test_multiple_issues(tmp_repo):
+    prepare_branch(tmp_repo, '10-11-sample-issue')
+
+    assert show_commit_message() == "Fix some issue\n\nIssue #10\n\nIssue #11\n\n"
+
+
+def test_text_before_issue(tmp_repo):
```
Comment (johndgiese) [inline tests/hook_test.py]: I think there are probably a few more test cases to handle.

E.g. merge commits

See if you can think of some more side cases, I just have a felling there are some.

File: tests/hook_test.py (hunk 6 above)
Comment (russellkan) [reply tests/hook_test.py]: For this part I came up with a few options:

1. `.strip()` the string and remove the `\n` from the assertion
2. Rearrange it and use `assert "text" in message()`
3. Use multi-line string for readability to format it as it would appear in the commit message

Which would you suggest I do?

File: rdm/cli.py (hunk 4 above)
Comment (russellkan) [reply rdm/cli.py]: I went ahead and renamed `src` and `dst` to `item_source` and `item_dest` and also changed `source` and `dest` to `dir_source` and `dir_dest` for clarity.

File: rdm/cli.py (hunk 3 above)
Comment (russellkan) [reply rdm/cli.py]: This makes sense; the change will be made in next push.

File: tests/install_hooks_test.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+@pytest.fixture
+def tmp_repo(tmpdir):
+    directory = str(tmpdir)
+    repo = Repo.init(directory)
+    os.chdir(directory)
+
+    yield repo
+
+    # teardown
```
Comment (johndgiese) [inline tests/install_hooks_test.py]: Do you think this is a worthwhile comment?

File: tests/install_hooks_test.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    subprocess.call(['rm', '-rf', directory])
+
+
+def test_install_hooks_no_destination(tmp_repo):
+    assert install_hooks(None) is None
```
Comment (johndgiese) [inline tests/install_hooks_test.py]: It seems like `install_hooks` always returns `None`.

See https://stackoverflow.com/questions/20274987/how-to-use-pytest-to-check-that-error-is-not-raised

I think the first suggestion in the answer is probably fine.

Including the asserts here is a bit misleading IMO

File: tests/hook_test.py (hunk 6 above)
Comment (johndgiese) [reply tests/hook_test.py]: I'm sorry; I actually meant the new lines in the tests themselves :).

I.e. use this:

```
def test_single_issue(tmp_repo):
    prepare_branch(tmp_repo, '10-sample-issue')
    assert show_commit_message() == "Fix some issue\n\nIssue #10\n\n"
```

I think the original assertions were probably better, since they were more precise.

File: tests/install_hooks_test.py (hunk 8 above)
Comment (russellkan) [reply tests/install_hooks_test.py]: I was using it just to keep track of the code flow when I was still testing how a fixture worked, but the teardown itself is pretty straightforward consisting of a single line so I'll delete it.

File: tests/hook_test.py (hunk 6 above)
Comment (russellkan) [reply tests/hook_test.py]: Oh okay, misunderstood that; I reverted the changes.
